explore/exploration.h

```diff
@@ -0,0 +1,180 @@
+#pragma once
+
+#include <stdint.h>
+#include <stdexcept>
+#include <algorithm>
+#include <numeric>
+#include <cstring>
+
+namespace exploration
+{
+  template<typename It>
+  void generate_epsilon_greedy(float epsilon, uint32_t top_action, It pdf_first, It pdf_last, std::random_access_iterator_tag pdf_tag)
+  {
+    size_t num_actions = pdf_last - pdf_first;
+    if (num_actions == 0)
+      return;
+
+	if (top_action >= num_actions)
+	#ifdef EXPLORE_NOEXCEPT
+  	  top_action = (uint32_t)num_actions - 1;
+  #else
+      throw std::out_of_range("top_action must be smaller than num_actions");
+  #endif
+
+    float prob = epsilon / (float)num_actions;
+
+    // size & initialize vector to prob
+    for (It d = pdf_first; d != pdf_last; ++d)
+      *d = prob;
+
+    *(pdf_first + top_action) += 1.f - epsilon;
+  }
+
+  template<typename It>
+  void generate_epsilon_greedy(float epsilon, uint32_t top_action, It pdf_first, It pdf_last)
+  {
+    typedef typename std::iterator_traits<It>::iterator_category pdf_category;
+    generate_epsilon_greedy(epsilon, top_action, pdf_first, pdf_last, pdf_category());
+  }
+
+  template<typename InputIt, typename OutputIt>
+  void generate_softmax(float lambda, InputIt scores_begin, InputIt scores_last, std::input_iterator_tag scores_tag, OutputIt pdf_first, OutputIt pdf_last, std::random_access_iterator_tag pdf_tag)
+  {
+    size_t num_actions_scores = scores_last - scores_begin;
```

Comment:
What happens if scores_last < scores_begin?

Reply:
Since scores_begin is an InputIterator it's not possible to check for scores_last <= scores_begin.